File: frontend/js/views/trade_expectancy.js
// Trade expectancy — per-trade edge from win rate + avg win/loss, the
// reward:risk ratio, break-even win rate, and expectancy in R, via
// /calc/trade-expectancy. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const FIELDS = [
    ['win_rate_pct', 'Win rate (%)', 40],
    ['avg_win_usd', 'Average win ($)', 300],
    ['avg_loss_usd', 'Average loss ($)', 100],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });

export async function renderTradeExpectancy(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.te.h1.title">// TRADE EXPECTANCY</span></h1>
        <p class="muted small" data-i18n="view.te.hint.intro">
            Whether a system makes money long-run isn't the win rate alone — it's the win rate
            weighted by how much you win vs lose: expectancy = (win% × avg win) − (loss% × avg
            loss). The break-even win rate is what you'd need at this reward:risk to net zero; a
            high win rate with a poor reward:risk can still lose. Expectancy in R is the per-trade
            edge in risk units. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.te.h2.inputs">Your stats</h2>
            <form id="te-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.te.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
        </div>
        <div id="te-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#te-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcTradeExpectancy(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.te.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#te-result');
    const edgeCls = r.has_edge ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.te.h2.result">The edge</h2>
            <div class="cards">
                <div class="card ${edgeCls}"><div class="label" data-i18n="view.te.card.expectancy">Expectancy / trade</div>
                    <div class="value ${edgeCls}">${money(r.expectancy_per_trade_usd)}</div></div>
                <div class="card ${edgeCls}"><div class="label" data-i18n="view.te.card.edge">Edge</div>
                    <div class="value ${edgeCls}">${r.has_edge ? t('view.te.yes') : t('view.te.no')}</div></div>
                <div class="card"><div class="label" data-i18n="view.te.card.rr">Reward : risk</div>
                    <div class="value">${Number(r.reward_risk_ratio).toFixed(2)} : 1</div></div>
                <div class="card"><div class="label" data-i18n="view.te.card.breakeven">Break-even win rate</div>
                    <div class="value">${Number(r.breakeven_win_rate_pct).toFixed(1)}%</div></div>
            </div>
            <table class="data-table">
                <thead><tr><th data-i18n="view.te.col.line">Line</th><th data-i18n="view.te.col.value">Value</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.te.row.r">Expectancy in R</td><td class="${edgeCls}">${Number(r.expectancy_in_r).toFixed(3)}R</td></tr>
                    <tr><td data-i18n="view.te.row.loss_rate">Loss rate</td><td>${Number(r.loss_rate_pct).toFixed(1)}%</td></tr>
                    <tr class="emph"><td data-i18n="view.te.row.per100">Expected over 100 trades</td><td class="${edgeCls}">${money(r.expectancy_per_100_trades_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/trade_expectancy.rs
//! Trade expectancy — the per-trade edge of a strategy.
//!
//! Whether a system makes money long-run isn't the win rate alone — it's the
//! win rate weighted by how much you win vs lose:
//!
//!   * expectancy = (win% × avg win) − (loss% × avg loss)
//!   * reward:risk = avg win / avg loss
//!   * break-even win rate = avg loss / (avg win + avg loss) — the win rate
//!     that makes expectancy zero at this reward:risk
//!   * expectancy in R = expectancy / avg loss (per-trade, in risk units)
//!
//! A positive expectancy is an edge; a high win rate with a poor reward:risk
//! can still lose. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ExpectancyInput {
    pub win_rate_pct: f64,
    pub avg_win_usd: f64,
    /// Average loss as a positive number.
    pub avg_loss_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct ExpectancyResult {
    pub loss_rate_pct: f64,
    /// Expected profit/loss per trade.
    pub expectancy_per_trade_usd: f64,
    pub reward_risk_ratio: f64,
    /// Win rate that yields zero expectancy at this reward:risk.
    pub breakeven_win_rate_pct: f64,
    /// Expectancy expressed in R (multiples of the average loss).
    pub expectancy_in_r: f64,
    /// Expected profit over 100 trades.
    pub expectancy_per_100_trades_usd: f64,
    /// True when expectancy is positive (a real edge).
    pub has_edge: bool,
}

pub fn analyze(i: &ExpectancyInput) -> ExpectancyResult {
    let win = (i.win_rate_pct / 100.0).clamp(0.0, 1.0);
    let loss_rate = 1.0 - win;
    let avg_win = i.avg_win_usd.max(0.0);
    let avg_loss = i.avg_loss_usd.max(0.0);

    let expectancy = win * avg_win - loss_rate * avg_loss;
    let reward_risk = if avg_loss > 0.0 { avg_win / avg_loss } else { 0.0 };
    let breakeven = if avg_win + avg_loss > 0.0 {
        avg_loss / (avg_win + avg_loss) * 100.0
    } else {
        0.0
    };
    let expectancy_r = if avg_loss > 0.0 { expectancy / avg_loss } else { 0.0 };

    ExpectancyResult {
        loss_rate_pct: loss_rate * 100.0,
        expectancy_per_trade_usd: expectancy,
        reward_risk_ratio: reward_risk,
        breakeven_win_rate_pct: breakeven,
        expectancy_in_r: expectancy_r,
        expectancy_per_100_trades_usd: expectancy * 100.0,
        has_edge: expectancy > 0.0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(win: f64, w: f64, l: f64) -> ExpectancyInput {
        ExpectancyInput { win_rate_pct: win, avg_win_usd: w, avg_loss_usd: l }
    }

    #[test]
    fn expectancy_formula() {
        // 40% win, win 300, loss 100: 0.4×300 − 0.6×100 = 120 − 60 = 60.
        let r = analyze(&inp(40.0, 300.0, 100.0));
        assert!((r.expectancy_per_trade_usd - 60.0).abs() < 1e-9);
        assert!((r.loss_rate_pct - 60.0).abs() < 1e-9);
    }

    #[test]
    fn reward_risk_ratio() {
        let r = analyze(&inp(40.0, 300.0, 100.0));
        assert!((r.reward_risk_ratio - 3.0).abs() < 1e-9);
    }

    #[test]
    fn breakeven_win_rate() {
        // loss/(win+loss) = 100/400 = 25%.
        let r = analyze(&inp(40.0, 300.0, 100.0));
        assert!((r.breakeven_win_rate_pct - 25.0).abs() < 1e-9);
    }

    #[test]
    fn has_edge_when_positive() {
        let r = analyze(&inp(40.0, 300.0, 100.0));
        assert!(r.has_edge); // 40% > 25% breakeven
    }

    #[test]
    fn high_win_rate_can_still_lose() {
        // 70% win but win 50, loss 200: 0.7×50 − 0.3×200 = 35 − 60 = −25.
        let r = analyze(&inp(70.0, 50.0, 200.0));
        assert!(r.expectancy_per_trade_usd < 0.0);
        assert!(!r.has_edge);
        // breakeven = 200/250 = 80% > 70% actual → loses.
        assert!((r.breakeven_win_rate_pct - 80.0).abs() < 1e-9);
    }

    #[test]
    fn at_breakeven_win_rate_expectancy_zero() {
        // R:R 3:1 → breakeven 25%; at exactly 25% expectancy ≈ 0.
        let r = analyze(&inp(25.0, 300.0, 100.0));
        assert!(r.expectancy_per_trade_usd.abs() < 1e-9);
    }

    #[test]
    fn expectancy_in_r_units() {
        // Expectancy 60 / avg loss 100 = 0.6R.
        let r = analyze(&inp(40.0, 300.0, 100.0));
        assert!((r.expectancy_in_r - 0.6).abs() < 1e-9);
        assert!((r.expectancy_per_100_trades_usd - 6_000.0).abs() < 1e-6);
    }

    #[test]
    fn higher_win_rate_raises_expectancy() {
        let low = analyze(&inp(30.0, 200.0, 100.0));
        let high = analyze(&inp(60.0, 200.0, 100.0));
        assert!(high.expectancy_per_trade_usd > low.expectancy_per_trade_usd);
    }
}
